### PR TITLE
sql/distsqlrun: refactor AlgebraicSetOp to implement RowSource

### LINKED ISSUE #21530: distsql: algebraicSetOp's EXCEPT ALL does not emit the correct number of duplicates
Page 276 of the SQL 1999 standard states that the number of duplicates of a row in the result of an `EXCEPT ALL` operation should be equal to `max(leftDuplicates - rightDuplicates, 0)`. I modified `algebraic_set_op_test.go` and added the following test case:
```go
...
		{
			spec: AlgebraicSetOpSpec{
				OpType: AlgebraicSetOpSpec_Except_all,
			},
			inputLeft: sqlbase.EncDatumRows{
				{v[2], v[3]},
				{v[5], v[6]},
				{v[2], v[3]},
				{v[5], v[6]},
				{v[2], v[6]},
				{v[3], v[5]},
				{v[2], v[9]},
			},
			inputRight: sqlbase.EncDatumRows{
				{v[2], v[3]},
				{v[5], v[6]},
				{v[5], v[6]},
			},
			expected: sqlbase.EncDatumRows{
				{v[2], v[3]},
				{v[2], v[6]},
				{v[3], v[5]},
				{v[2], v[9]},
			},
		},
...
```

(Note that `v[i] == i`.)

However, I get the following error: `invalid results: [[2 6] [3 5] [2 9]], expected [[2 3] [2 6] [3 5] [2 9]]`.

Instead of emitting one `[2 3]` (2 left duplicates - 1 right duplicate), we emit 0 copies of it.

### PR DESCRIPTION
**Pull request by rjnn**
Fixes #21530. 

@abhimadan: you can probably rebase over this and make progress on planning immediately. You can rebase over master once it merges. I'll help you if the review process makes any breaking changes.

### CODE REVIEW

File: pkg/sql/distsqlrun/algebraic_set_op.go [hunk 1]
```diff
@@ -37,9 +35,20 @@ type algebraicSetOp struct {
 	opType                  AlgebraicSetOpSpec_SetOpType
 	ordering                Ordering
 	types                   []sqlbase.ColumnType
-	datumAlloc              *sqlbase.DatumAlloc
+	datumAlloc              sqlbase.DatumAlloc
+
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/algebraic_set_op.go]: I assume this break means that everything below is runtime state - could you please add a comment that makes this explicit? We started doing this explicitly (with a separate struct - but even a comment will be helpful) in planNodes and it makes understanding the code a lot more straightforward.

File: pkg/sql/distsqlrun/algebraic_set_op.go [hunk 2]
```diff
@@ -37,9 +35,20 @@ type algebraicSetOp struct {
...
+	leftAccum, rightAccum streamGroupAccumulator
+	leftGroup, rightGroup []sqlbase.EncDatumRow
+	allCols               columns
+	rightMap              map[string]int
+	leftIdx               int
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/algebraic_set_op.go]: Could you document some of these fields? At least the ones that don't have obvious meanings (to me anyway) like cmp, leftIdx, rightMap.

File: pkg/sql/distsqlrun/algebraic_set_op.go [hunk 3]
```diff
@@ -88,183 +98,250 @@ func newAlgebraicSetOp(
...
-	defer e.rightSource.ConsumerDone()
+func (e *algebraicSetOp) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
+	var row sqlbase.EncDatumRow
+	var meta *ProducerMetadata
 
-	e.evalCtx = e.flowCtx.NewEvalCtx()
+	for {
+		switch e.opType {
+		case AlgebraicSetOpSpec_Except_all:
+			row, meta = e.nextExceptAll()
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/algebraic_set_op.go]: It seems to me that this is a relatively easy opportunity to static-ify the responsibility of this processor. Can we define an interface for algebraic set ops (doesn't have to be a Go interface, just like a protocol) and then plug the correct one into this processor at planning time? That way, we won't need to perform this switch on every row. I'm not trying to suggest you do that in this PR- just noticed this and writing it down.

**Review by jordanlewis — COMMENTED** [review]
I didn't carefully review the algorithm, but this generally LGTM. I left a couple of comments.

File: pkg/sql/distsqlrun/algebraic_set_op.go [hunk 4]
```diff
@@ -88,183 +98,250 @@ func newAlgebraicSetOp(
...
-	default:
-		panic(fmt.Sprintf("cannot run unsupported algebraicSetOp %v", e.opType))
+		outRow, status, err := e.out.ProcessRow(e.ctx, row)
+		if err != nil {
+			return nil, e.producerMeta(err)
+		}
+		switch status {
+		case NeedMoreRows:
+			if outRow == nil && err == nil {
+				log.Infof(e.evalCtx.Ctx(), "outrow is nil, but need more rows. skipping")
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/algebraic_set_op.go]: Not familiar with logging in Distsql land, but I would suggest putting this behind a V(2) level log if this is just an ordinary Infof.

File: pkg/sql/distsqlrun/algebraic_set_op.go [hunk 5]
```diff
@@ -88,183 +98,250 @@ func newAlgebraicSetOp(
...
+				continue
+			}
+		case DrainRequested:
+			e.leftSource.ConsumerDone()
+			e.rightSource.ConsumerDone()
+			log.Infof(e.evalCtx.Ctx(), "drain requested. skipping")
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/algebraic_set_op.go]: Ditto

File: pkg/sql/distsqlrun/algebraic_set_op.go [hunk 6]
```diff
@@ -88,183 +98,250 @@ func newAlgebraicSetOp(
...
-				return err
+// advanceLeft advances the left group. The error field must be checked.
+func (e *algebraicSetOp) advanceLeft() {
+	e.leftGroup, e.err = e.leftAccum.advanceGroup(e.evalCtx, e.out.output)
+	if e.leftGroup != nil {
+		e.resetIndexes()
+	}
+}
+
+func (e *algebraicSetOp) advanceRight() {
```
Comment (jordanlewis) [inline pkg/sql/distsqlrun/algebraic_set_op.go]: Please add a comment for consistency with advanceLeft()

**Comment by abhimadan** [comment]
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 296 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3AVRn8uxaPyBztX_j-:-L3AVRn8uxaPyBztX_j0:b2v28nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L296)):*
> ```Go
> func (e *algebraicSetOp) advanceRight() {
> 	e.rightGroup, e.err = e.rightAccum.advanceGroup(e.evalCtx, e.out.output)
> 	e.rightMap = nil
> ```

Might just be my lack of understanding here, but is it correct to clear the map if there's no ordering? In that case, we can't guarantee that rows earlier in the right-side stream won't appear later in the left-side stream. The tests probably won't catch it it since it would only become a problem if the stream will be split into multiple stream groups, and none of them seem to make tables with more than 64 rows.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 330 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3AT4uMuBjSbWItZ40p:-L3AT4uNjYyFbw3Q7zZu:b-s2w66i) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L330)):*
> ```Go
> 				return
> 			}
> 			if _, ok := e.rightMap[string(encoded)]; ok {
> ```

I think you can get rid of this if statement altogether and just do `rightMap[string(encoded)]++`.

---

*[pkg/sql/distsqlrun/algebraic_set_op_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ATTW8ofd4fwmk3EXE:-L3ATTW8ofd4fwmk3EXF:bxmib4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op_test.go#L198)):*
> ```Go
> 				OpType: AlgebraicSetOpSpec_Except_all,
> 			},
> 			inputLeft: sqlbase.EncDatumRows{
> ```

nit: you can replace `inputLeft` with `td.inputUnordered`, since I just copied it to make the bug report more complete.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560)*

**Comment by arjunravinarayan** [comment]
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ADGxWrqDPCwyCIXFr-r1-39:-L3EJjUjEPCW3pbhVJub:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L39)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I assume this break means that everything below is runtime state - could you please add a comment that makes this explicit? We started doing this explicitly (with a separate struct - but even a comment will be helpful) in planNodes and it makes understanding the code a lot more straightforward.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ADGxWrqDPCwyCIXFr-r1-44:-L3ELJ4rLjI5hpEv6chq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L44)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Could you document some of these fields? At least the ones that don't have obvious meanings (to me anyway) like cmp, leftIdx, rightMap.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ADGxWrqDPCwyCIXFr-r1-136:-L3EQ99dsten0sbOPXzC:b-z0k87d) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L136)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It seems to me that this is a relatively easy opportunity to static-ify the responsibility of this processor. Can we define an interface for algebraic set ops (doesn't have to be a Go interface, just like a protocol) and then plug the correct one into this processor at planning time? That way, we won't need to perform this switch on every row. I'm not trying to suggest you do that in this PR- just noticed this and writing it down.
</blockquote></details>

I think this should just be an `ExceptAll` processor, honestly. Will change.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ADGxWrqDPCwyCIXFr-r1-155:-L3EMr-HV3nzByA46KoS:b-x9eacn) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L155)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Not familiar with logging in Distsql land, but I would suggest putting this behind a V(2) level log if this is just an ordinary Infof.
</blockquote></details>

I think these logs should just go. Removed.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ADGxWrqDPCwyCIXFr-r1-161:-L3EMxESDDO-MrM50dwh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L161)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 294 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ADGxWrqDPCwyCIXFr-r1-294:-L3EN2Gw5bkkC5mPosjE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L294)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Please add a comment for consistency with advanceLeft()
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 296 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3AVRn8uxaPyBztX_j-:-L3EN8jHjFMrF_HChkzq:b-p4e2b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L296)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

Might just be my lack of understanding here, but is it correct to clear the map if there's no ordering? In that case, we can't guarantee that rows earlier in the right-side stream won't appear later in the left-side stream. The tests probably won't catch it it since it would only become a problem if the stream will be split into multiple stream groups, and none of them seem to make tables with more than 64 rows.
</blockquote></details>

If there's no ordering, then everything is in a single group. Then this function will never be called. 

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 330 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3AT4uMuBjSbWItZ40p:-L3ENMd0C-Fx0IhB4DVE:bm0mm0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op.go#L330)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

I think you can get rid of this if statement altogether and just do `rightMap[string(encoded)]++`.
</blockquote></details>

You're right! Zero values of maps, TIL.

---

*[pkg/sql/distsqlrun/algebraic_set_op_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3ATTW8ofd4fwmk3EXE:-L3EOFFnK9nnRHiyVSaw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11c252efe1d4265d4ee6a63f7d5ec4009f676b9f/pkg/sql/distsqlrun/algebraic_set_op_test.go#L198)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

nit: you can replace `inputLeft` with `td.inputUnordered`, since I just copied it to make the bug report more complete.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560)*

**Comment by petermattis** [comment]
It is worthwhile to add a benchmark for `algebraicSetOp` similar to the other processor benchmarks (e.g. `BenchmarkGrouping`). I found when converting other processors to implement `RowSource` that it was pretty easy to make a simple change that hurt performance.

Should `algebraicSetOp` be renamed `exceptAll`? Fine to do in a follow-on PR so that this one doesn't get muddled.

Is `EXCEPT ALL` the same as semi-join where the join condition is all of the columns?

---

Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3E_43HzlQdBO94cGlw:-L3E_43HzlQdBO94cGlx:b-pia3w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L42)):*
> ```Go
> 	// The following variables retain the state of the processor between
> 	// successive calls to Next().
> 	// leftGroup and righ
> ```

You've got an incomplete sentence here.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 172 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EahDuTjQN697VjSnB:-L3EahDuTjQN697VjSnC:b-wk1am0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L172)):*
> ```Go
> }
> 
> // nextExceptAll pushes all rows in the left stream that are not present in the
> ```

This comment needs updating. This method no longer "pushes".

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 334 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EaMlz5gQZMZFthbZs:-L3EaMlz5gQZMZFthbZt:bl5ipgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L334)):*
> ```Go
> 				return
> 			}
> 			e.rightMap[string(encoded)]++
> ```

You might want to use `util/stringarena` here. I bet a benchmark would show a boost.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3E_I98Jrkve4vTE0OF:b-pkewvz)*

**Comment by abhimadan** [comment]
Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 47 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EUT9nX0iWeoxdWTv7:-L3EUT9nX0iWeoxdWTv8:bpkeqls) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L47)):*
> ```Go
> 	// in leftGroup.
> 	rightMap              map[string]int
> 	// leftIdx is the index into leftGroup where we resume iteration
> ```

Please add a period

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 160 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EVeRglIAF-7LoPogg:-L3EVeRglIAF-7LoPogh:bhgix7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L160)):*
> ```Go
> 		switch status {
> 		case NeedMoreRows:
> 			if outRow == nil && err == nil {
> ```

Since you already know that `err == nil` here, you can remove the check inside the case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560)*

**Comment by petermattis** [comment]
> Is EXCEPT ALL the same as semi-join where the join condition is all of the columns?

https://technet.microsoft.com/en-us/library/ms191255(v=sql.105).aspx

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3EgWj70ZsY8ZiNHCUY:b-5ycxwf)*

**Comment by petermattis** [comment]
Err, make that `EXCEPT ALL` is a left anti semi join. One advantage of using the join code for this is that we'll automatically take advantage of the disk buffering if necessary. Right now, `algebraicSetOp` has no memory monitoring or disk buffering (AFAICT).

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3Eh_1Q9pyVeDqCnGeq:bavqp1e)*

**Comment by abhimadan** [comment]
How do semi-joins and anti-joins deal with duplicates? From the quick reading I did, it seems like they only return one copy of each duplicate. It seems like using them would be a win for `EXCEPT` and `INTERSECT` (vs. `{EXCEPT, INTERSECT} ALL` + `DISTINCT`), but it may not produce the correct behaviour when we want to keep duplicates in the result table.

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3EibI3iKhtmjmy9Q4f:bsagoaq)*

**Comment by arjunravinarayan** [comment]
@petermattis see #21530

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3Ejy-_huNaWPCF49uM:b-6uanle)*

**Comment by petermattis** [comment]
The duplicate behavior might need slight adjustments between {semi,anti}-join and {intersect,except}, but I don't think that is a reason for a fully separate processor, just an area to pay attention to. 

I think this PR is fine to merge, but let's file an issue about using semi/anti join instead so we can avoid maintaining/enhancing it soon.

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3Et85jkqrGHVlhRNzt:bd8d7d0)*

**Comment by arjunravinarayan** [comment]
I've filed an issue (#21706). Adding benchmarks now.

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3E_43HzlQdBO94cGlw:-L3Ym357RzWvybGgefWH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L42)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You've got an incomplete sentence here.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 47 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EUT9nX0iWeoxdWTv7:-L3Ym5qj74MHLln05A6f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L47)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

Please add a period
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 160 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EVeRglIAF-7LoPogg:-L3YmfL1Ez0nShSr500s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L160)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

Since you already know that `err == nil` here, you can remove the check inside the case.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/algebraic_set_op.go, line 172 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-L3EahDuTjQN697VjSnB:-L3YmmvhMx_JEzoVDFJz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf76b871b7cc06a875f30316985afce78983241f/pkg/sql/distsqlrun/algebraic_set_op.go#L172)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This comment needs updating. This method no longer "pushes".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3YnNR_JodsZZQ1PBz-:b8ovjtb)*

**Comment by arjunravinarayan** [comment]
Prepended a commit that adds a benchmark. Investigating squeezing out a little more performance with `stringarena` now.

```name         old time/op    new time/op    delta
ExceptAll-8     548µs ± 3%     429µs ± 4%  -21.67%  (p=0.000 n=10+9)

name         old speed      new speed      delta
ExceptAll-8  43.8MB/s ± 3%  55.9MB/s ± 4%  +27.73%  (p=0.000 n=10+9)

name         old alloc/op   new alloc/op   delta
ExceptAll-8     354kB ± 0%     320kB ± 0%   -9.59%  (p=0.000 n=8+10)

name         old allocs/op  new allocs/op  delta
ExceptAll-8     3.08k ± 0%     1.08k ± 0%  -64.83%  (p=0.000 n=10+10)
```

---

Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3YtJ6LPqgSY0s_DfXE:b-cuykx5)*

**Comment by petermattis** [comment]
:lgtm:

> Prepended a commit that adds a benchmark. Investigating squeezing out a little more performance with stringarena now.

Do you think squeezing extra performance here is warranted if we're going to try to get rid of `algebraicSetOp` soon?

---

Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3Z27DqXx9ESwoWawpT:b-gs4n14)*

**Comment by arjunravinarayan** [comment]
You're right. Merging on green and moving on to other processors.

---

Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21560#-:-L3ZJEFnQ2HcAvEto714:b-533fxn)*

**Pull request merged**